### PR TITLE
[mojo-stdlib] Make assert_equal more generic by using traits

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
This should help write tests with new structs like `pathlib.Path`, `Bytes`, datetime, etc...

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
I'm not sure I understand why the `mojo format` is failing in the CI. Locally, here is what I get:
```
$ modular update nightly/mojo && mojo format stdlib/src/testing/testing.mojo
# No updates available for this packages found.
All done! ✨ 🍰 ✨
1 file left unchanged.
```

**Comment by ConnorGray** [comment]
Hi Gabriel, the formatting CI is failing because it runs with the `-l 80` flag that causes the formatter to limit lines to 80 columns in width:

```
mojo format stdlib/src/testing/testing.mojo -l 80
```

If you add `-l 80` to your local invocation, it should format at the CI expects.

File: stdlib/src/testing/testing.mojo [hunk 1]
```diff
@@ -109,6 +114,7 @@ fn assert_equal(lhs: Int, rhs: Int, msg: String = "") raises:
         raise _assert_equal_error(str(lhs), str(rhs), msg=msg)
 
 
+# TODO: Remove the String and SIMD overloads once we have more powerful traits.
```
Comment (abduld) [inline stdlib/src/testing/testing.mojo:117]: why can't you remove these now?

**Review by abduld — APPROVED** [review]
looks great. Thanks

Can you add a test using your favorite type as well

**Comment by gabrieldemarmiesse** [comment]
> Can you add a test using your favorite type as well

Done :)

**Comment by ConnorGray** [comment]
Hey Gabriel, this looks good, thanks for the contribution and adding those tests! :)

Going to proceed with merging this now.

**Pull request merged**